### ISSUE TITLE
Live subscribe/unsubscribe for Binance exchanges

### ISSUE THREAD

**Issue by bmancini55** [altangent/ccxws#161]
Issue initially reported in PR #152.  Binance supports live subscribe/unsubscribe - https://github.com/binance-exchange/binance-official-api-docs/blob/master/web-socket-streams.md#live-subscribingunsubscribing-to-streams.  It is likely that the Binance Clients can be migrated to use the `BasicClient` instead of custom implementations.

**Comment by bmancini55** [comment]
Looks like a socket can be initialized without an attached stream:
```javascript
const ws = new WebSocket("wss://stream.binance.com:9443/stream");
ws.onopen = function() {
    console.log("open");
}
ws.onmessage = function(msg) {
    const data = JSON.parse(msg.data);
    console.log(data.stream);
}

ws.send(JSON.stringify(
        {
          method: "SUBSCRIBE",
          params: ['btcusdt@aggTrade', 'ethbtc@aggTrade'],
          id: 1
        }
    ))

ws.send(JSON.stringify({
    method: "UNSUBSCRIBE",
    params: ['btcusdt@aggTrade'],
    id: 1
}))
```

**Comment by bmancini55** [comment]
It looks like there there is a limit to the number of sub/unsub operations that can be performed in a given time period. I've tested this with a 1-sub semaphore and it works decently with 500ms delays per sub. 

For example, this will result in 0 subscriptions:
```
const ws = new WebSocket("wss://stream.binance.com:9443/stream");
ws.onopen = function() {
    console.log("open");
}
ws.onmessage = function(msg) {
    const data = JSON.parse(msg.data);
    console.log(data.stream ? data.stream : data);
}

ws.send(JSON.stringify({
  method: 'SUBSCRIBE',
  params: ["ethbtc@aggTrade"],
  id: 0
}))

ws.send(JSON.stringify({
  method: 'SUBSCRIBE',
  params: ["ltcbtc@aggTrade"],
  id: 1
}))

ws.send(JSON.stringify({
  method: 'SUBSCRIBE',
  params: ["bnbbtc@aggTrade"],
  id: 2
}))

ws.send(JSON.stringify({
  method: 'SUBSCRIBE',
  params: ["neobtc@aggTrade"],
  id: 3
}))

ws.send(JSON.stringify({
  method: 'SUBSCRIBE',
  params: ["qtumeth@aggTrade"],
  id: 4
}))

ws.send(JSON.stringify({
  method: 'SUBSCRIBE',
  params: ["eoseth@aggTrade"],
  id: 5
}))

ws.send(JSON.stringify({
  method: 'SUBSCRIBE',
  params: ["snteth@aggTrade"],
  id: 6
}))

ws.send(JSON.stringify({
  method: 'SUBSCRIBE',
  params: ["bnteth@aggTrade"],
  id: 7
}))

ws.send(JSON.stringify({
  method: 'SUBSCRIBE',
  params: ["gasbtc@aggTrade"],
  id: 8
}))

ws.send(JSON.stringify({
  method: 'SUBSCRIBE',
  params: ["bnbeth@aggTrade"],
  id: 9
}))
```

**Comment by bmancini55** [comment]
There is also a limit to the number of subscriptions that can be made at at time.  I'm getting a failure with 437 streams, though this could easily be param field size limit, binance silently fails without any additional information.  

For example, this fails, but if you try again with 1 less stream it will work:

```
const ws = new WebSocket("wss://stream.binance.com:9443/stream");
ws.onopen = function() {
    console.log("open");
}
ws.onmessage = function(msg) {
    const data = JSON.parse(msg.data);
    console.log(data.stream ? data.stream : data);
}

ws.send(JSON.stringify({
  method: 'SUBSCRIBE',
  params: ["ethbtc@aggTrade","ltcbtc@aggTrade","bnbbtc@aggTrade","neobtc@aggTrade","qtumeth@aggTrade","eoseth@aggTrade","snteth@aggTrade","bnteth@aggTrade","gasbtc@aggTrade","bnbeth@aggTrade","btcusdt@aggTrade","ethusdt@aggTrade","mcoeth@aggTrade","mcobtc@aggTrade","wtcbtc@aggTrade","wtceth@aggTrade","lrcbtc@aggTrade","lrceth@aggTrade","qtumbtc@aggTrade","yoyobtc@aggTrade","omgbtc@aggTrade","omgeth@aggTrade","zrxbtc@aggTrade","zrxeth@aggTrade","stratbtc@aggTrade","strateth@aggTrade","snglsbtc@aggTrade","bqxbtc@aggTrade","bqxeth@aggTrade","kncbtc@aggTrade","knceth@aggTrade","funbtc@aggTrade","funeth@aggTrade","snmbtc@aggTrade","neoeth@aggTrade","iotabtc@aggTrade","iotaeth@aggTrade","linkbtc@aggTrade","linketh@aggTrade","xvgbtc@aggTrade","xvgeth@aggTrade","mdabtc@aggTrade","mdaeth@aggTrade","mtlbtc@aggTrade","mtleth@aggTrade","eosbtc@aggTrade","sntbtc@aggTrade","etceth@aggTrade","etcbtc@aggTrade","mthbtc@aggTrade","engbtc@aggTrade","engeth@aggTrade","dntbtc@aggTrade","zecbtc@aggTrade","zeceth@aggTrade","bntbtc@aggTrade","astbtc@aggTrade","dashbtc@aggTrade","dasheth@aggTrade","oaxbtc@aggTrade","btgbtc@aggTrade","evxbtc@aggTrade","evxeth@aggTrade","reqbtc@aggTrade","vibbtc@aggTrade","vibeth@aggTrade","trxbtc@aggTrade","trxeth@aggTrade","powrbtc@aggTrade","powreth@aggTrade","arkbtc@aggTrade","arketh@aggTrade","xrpbtc@aggTrade","xrpeth@aggTrade","enjbtc@aggTrade","enjeth@aggTrade","storjbtc@aggTrade","storjeth@aggTrade","bnbusdt@aggTrade","kmdbtc@aggTrade","kmdeth@aggTrade","rcnbtc@aggTrade","nulsbtc@aggTrade","nulseth@aggTrade","rdnbtc@aggTrade","rdneth@aggTrade","xmrbtc@aggTrade","xmreth@aggTrade","wtcbnb@aggTrade","dltbtc@aggTrade","ambbtc@aggTrade","batbtc@aggTrade","bateth@aggTrade","batbnb@aggTrade","bcptbtc@aggTrade","arnbtc@aggTrade","arneth@aggTrade","gvtbtc@aggTrade","cdtbtc@aggTrade","cdteth@aggTrade","gxsbtc@aggTrade","gxseth@aggTrade","neousdt@aggTrade","neobnb@aggTrade","poebtc@aggTrade","qspbtc@aggTrade","qspeth@aggTrade","btsbtc@aggTrade","xzcbtc@aggTrade","xzceth@aggTrade","xzcbnb@aggTrade","lskbtc@aggTrade","lsketh@aggTrade","tntbtc@aggTrade","tnteth@aggTrade","fuelbtc@aggTrade","manabtc@aggTrade","manaeth@aggTrade","bcdbtc@aggTrade","bcdeth@aggTrade","iotabnb@aggTrade","adxbtc@aggTrade","adxeth@aggTrade","adabtc@aggTrade","adaeth@aggTrade","pptbtc@aggTrade","cmtbtc@aggTrade","cmteth@aggTrade","xlmbtc@aggTrade","xlmeth@aggTrade","xlmbnb@aggTrade","cndbtc@aggTrade","lendbtc@aggTrade","lendeth@aggTrade","wabibtc@aggTrade","wabibnb@aggTrade","ltceth@aggTrade","ltcusdt@aggTrade","ltcbnb@aggTrade","tnbbtc@aggTrade","wavesbtc@aggTrade","waveseth@aggTrade","wavesbnb@aggTrade","gtobtc@aggTrade","gtoeth@aggTrade","icxbtc@aggTrade","icxeth@aggTrade","icxbnb@aggTrade","ostbtc@aggTrade","osteth@aggTrade","elfbtc@aggTrade","elfeth@aggTrade","aionbtc@aggTrade","aioneth@aggTrade","aionbnb@aggTrade","neblbtc@aggTrade","nebleth@aggTrade","neblbnb@aggTrade","brdbtc@aggTrade","brdeth@aggTrade","brdbnb@aggTrade","edobtc@aggTrade","edoeth@aggTrade","navbtc@aggTrade","lunbtc@aggTrade","appcbtc@aggTrade","vibebtc@aggTrade","rlcbtc@aggTrade","rlceth@aggTrade","rlcbnb@aggTrade","insbtc@aggTrade","pivxbtc@aggTrade","pivxeth@aggTrade","iostbtc@aggTrade","iosteth@aggTrade","steembtc@aggTrade","steemeth@aggTrade","steembnb@aggTrade","nanobtc@aggTrade","nanoeth@aggTrade","nanobnb@aggTrade","viabtc@aggTrade","blzbtc@aggTrade","blzeth@aggTrade","blzbnb@aggTrade","aebtc@aggTrade","aeeth@aggTrade","ncasheth@aggTrade","poabtc@aggTrade","zilbtc@aggTrade","zileth@aggTrade","zilbnb@aggTrade","ontbtc@aggTrade","onteth@aggTrade","ontbnb@aggTrade","stormbtc@aggTrade","stormeth@aggTrade","qtumusdt@aggTrade","xembtc@aggTrade","xemeth@aggTrade","wanbtc@aggTrade","waneth@aggTrade","wanbnb@aggTrade","wprbtc@aggTrade","qlcbtc@aggTrade","qlceth@aggTrade","sysbtc@aggTrade","grsbtc@aggTrade","grseth@aggTrade","adausdt@aggTrade","adabnb@aggTrade","gntbtc@aggTrade","gnteth@aggTrade","loombtc@aggTrade","loometh@aggTrade","xrpusdt@aggTrade","repbtc@aggTrade","repeth@aggTrade","btctusd@aggTrade","ethtusd@aggTrade","zenbtc@aggTrade","zeneth@aggTrade","zenbnb@aggTrade","skybtc@aggTrade","skyeth@aggTrade","eosusdt@aggTrade","eosbnb@aggTrade","cvcbtc@aggTrade","cvceth@aggTrade","thetabtc@aggTrade","thetaeth@aggTrade","thetabnb@aggTrade","xrpbnb@aggTrade","tusdusdt@aggTrade","iotausdt@aggTrade","xlmusdt@aggTrade","iotxbtc@aggTrade","iotxeth@aggTrade","qkcbtc@aggTrade","qkceth@aggTrade","agibtc@aggTrade","agieth@aggTrade","nxsbtc@aggTrade","enjbnb@aggTrade","databtc@aggTrade","dataeth@aggTrade","ontusdt@aggTrade","trxbnb@aggTrade","trxusdt@aggTrade","etcusdt@aggTrade","etcbnb@aggTrade","icxusdt@aggTrade","scbtc@aggTrade","sceth@aggTrade","scbnb@aggTrade","npxseth@aggTrade","keyeth@aggTrade","nasbtc@aggTrade","naseth@aggTrade","mfteth@aggTrade","mftbnb@aggTrade","denteth@aggTrade","ardrbtc@aggTrade","nulsusdt@aggTrade","hotbtc@aggTrade","hoteth@aggTrade","vetbtc@aggTrade","veteth@aggTrade","vetusdt@aggTrade","vetbnb@aggTrade","dockbtc@aggTrade","docketh@aggTrade","polybtc@aggTrade","hcbtc@aggTrade","gobtc@aggTrade","paxusdt@aggTrade","rvnbtc@aggTrade","rvnbnb@aggTrade","dcrbtc@aggTrade","mithbtc@aggTrade","mithbnb@aggTrade","bnbpax@aggTrade","btcpax@aggTrade","ethpax@aggTrade","xrppax@aggTrade","eospax@aggTrade","renbtc@aggTrade","bnbtusd@aggTrade","xrptusd@aggTrade","eostusd@aggTrade","xlmtusd@aggTrade","bnbusdc@aggTrade","btcusdc@aggTrade","ethusdc@aggTrade","xrpusdc@aggTrade","eosusdc@aggTrade","usdcusdt@aggTrade","adatusd@aggTrade","trxtusd@aggTrade","neotusd@aggTrade","trxxrp@aggTrade","paxtusd@aggTrade","usdctusd@aggTrade","usdcpax@aggTrade","linkusdt@aggTrade","linktusd@aggTrade","linkpax@aggTrade","linkusdc@aggTrade","wavesusdt@aggTrade","wavesusdc@aggTrade","ltctusd@aggTrade","ltcpax@aggTrade","ltcusdc@aggTrade","trxpax@aggTrade","trxusdc@aggTrade","bttbnb@aggTrade","bttusdt@aggTrade","usdsusdt@aggTrade","bttpax@aggTrade","btttusd@aggTrade","bttusdc@aggTrade","ongbtc@aggTrade","ongusdt@aggTrade","hotbnb@aggTrade","hotusdt@aggTrade","zilusdt@aggTrade","zrxbnb@aggTrade","zrxusdt@aggTrade","fetbnb@aggTrade","fetbtc@aggTrade","fetusdt@aggTrade","batusdt@aggTrade","xmrbnb@aggTrade","xmrusdt@aggTrade","zecbnb@aggTrade","zecusdt@aggTrade","zectusd@aggTrade","zecusdc@aggTrade","iostbnb@aggTrade","iostusdt@aggTrade","celrbnb@aggTrade","celrbtc@aggTrade","celrusdt@aggTrade","adausdc@aggTrade","neousdc@aggTrade","dashbnb@aggTrade","dashusdt@aggTrade","nanousdt@aggTrade","omgusdt@aggTrade","thetausdt@aggTrade","enjusdt@aggTrade","mithusdt@aggTrade","maticbnb@aggTrade","maticbtc@aggTrade","maticusdt@aggTrade","atombnb@aggTrade","atombtc@aggTrade","atomusdt@aggTrade","atomusdc@aggTrade","batusdc@aggTrade","phbbtc@aggTrade","phbtusd@aggTrade","tfuelbtc@aggTrade","tfuelusdt@aggTrade","onebnb@aggTrade","onebtc@aggTrade","oneusdt@aggTrade","oneusdc@aggTrade","ftmbnb@aggTrade","ftmbtc@aggTrade","ftmusdt@aggTrade","algobnb@aggTrade","algobtc@aggTrade","algousdt@aggTrade","algotusd@aggTrade","algopax@aggTrade","gtousdt@aggTrade","erdbnb@aggTrade","erdbtc@aggTrade","erdusdt@aggTrade","dogebnb@aggTrade","dogebtc@aggTrade","dogeusdt@aggTrade","duskbtc@aggTrade","duskusdt@aggTrade","bgbpusdc@aggTrade","ankrbnb@aggTrade","ankrbtc@aggTrade","ankrusdt@aggTrade","winbnb@aggTrade","winusdt@aggTrade","winusdc@aggTrade","cosbnb@aggTrade","cosbtc@aggTrade","cosusdt@aggTrade","npxsusdt@aggTrade","cocosbnb@aggTrade","cocosusdt@aggTrade","mtlusdt@aggTrade","tomobnb@aggTrade","tomobtc@aggTrade","tomousdt@aggTrade","perlbnb@aggTrade","perlbtc@aggTrade","perlusdt@aggTrade","dentusdt@aggTrade","mftusdt@aggTrade","keyusdt@aggTrade","stormusdt@aggTrade","dockusdt@aggTrade","wanusdt@aggTrade","funusdt@aggTrade","cvcusdt@aggTrade","btttrx@aggTrade","wintrx@aggTrade","chzbnb@aggTrade","chzbtc@aggTrade","chzusdt@aggTrade","bandbnb@aggTrade","bandbtc@aggTrade","bandusdt@aggTrade","bnbbusd@aggTrade","btcbusd@aggTrade","busdusdt@aggTrade","beambnb@aggTrade","beambtc@aggTrade","beamusdt@aggTrade","xtzbnb@aggTrade","xtzbtc@aggTrade","xtzusdt@aggTrade","renusdt@aggTrade","rvnusdt@aggTrade","hcusdt@aggTrade","hbarbnb@aggTrade","hbarbtc@aggTrade","hbarusdt@aggTrade","nknbnb@aggTrade","nknbtc@aggTrade"],
  id: 1
}))

```

**Comment by FishTheOriginal** [comment]
I can confirm its a query length limit from testing. Binance throws a 414 when its unhappy way.
`Error: Unexpected server response: 414`

**Comment by bmancini55** [comment]
@FishTheOriginal thanks I added both throttling and batching which seems to stabilize things well

**Comment by FishTheOriginal** [comment]
I would like to thank @bmancini55 and everyone who was involved in this issue. Now not to ruin the party there's an issue with handling errors.

Errors are not being handled:

`TypeError: Cannot read property 'toLowerCase' of undefined
    at BinanceClient._onMessage (/.../node_modules/ccxws/src/exchanges/binance-base.js:208:20)`

What happens is that **msg** looks like this:
`{ error: { code: 4, msg: 'Too many subscriptions' }, id: 6 }`

**Comment by bmancini55** [comment]
@FishTheOriginal thanks the heads up!  Looks like an easy fix that I'll get in shortly.

**Comment by bmancini55** [comment]
Fixed in new release v0.34.1!

**Comment by FishTheOriginal** [comment]
I'm putting some stress on Binance and the lib seems to be holding well. Thanks @bmancini55 .
